modules/custom/az_migration/az_migration_remote/src/Plugin/migrate/process/MigrationRemoteFile.php

```diff
@@ -0,0 +1,278 @@
+    $instance->pluginManagerMigration = $container->get('plugin.manager.migration');
+    return $instance;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function transform($value, MigrateExecutableInterface $migrate_executable, Row $row, $destination_property) {
+
+    // Url is provided as the value of the process plugin.
+    $url = $value;
+    if (is_array($url)) {
+      $url = reset($url);
+    }
+    $original_uri = NULL;
+
+    // Lookup the existing uri. The existing uri constitutes our default.
+    // We will return this if unchanged or can't fetch the new file.
+    $migration = $this->configuration['migration'] ?? NULL;
+    $source_ids = $this->configuration['source_ids'] ?? [];
+    $source_ids = $row->getMultiple($source_ids);
+    if (!empty($migration) && !empty($source_ids)) {
+      // Lookup the file migration to find the current file.
+      try {
+        $destination_id_array = $this->migrateLookup->lookup($migration, $source_ids);
+      }
+      catch (\Exception $e) {
+        $destination_id_array = [];
+      }
+      $destination_id_array = reset($destination_id_array);
+      if (!empty($destination_id_array)) {
+        // Attempt to load the destination file entity and see if it has a uri.
+        $file = $this->entityTypeManager->getStorage('file')->load(reset($destination_id_array));
+        if (!empty($file)) {
+          $original_uri = $file->getFileUri();
+        }
+      }
+    }
+
+    // There's nothing to fetch if we don't have an URL.
+    if (empty($url)) {
+      return $original_uri;
+    }
+
+    try {
+      // Request the remote resource.
+      $response = $this->httpClient->get($url, [
+        // Specify an accept header to indicate what we're interested in.
+        'headers'        => ['Accept' => static::REQUEST_ACCEPT],
+      ]);
+    }
+    catch (\Exception $e) {
+      $this->logger->notice("During fetch of %url, %msg", [
+        '%url' => $url,
+        '%msg' => $e->getMessage(),
+      ]);
+      return $original_uri;
+    }
+    // Get headers relevant to file.
+    $type = $response->getHeader('Content-Type') ?? [];
+    $type = array_pop($type);
+    if (empty($type)) {
+      $this->logger->notice("No Content-Type provided by %url", [
+        '%url' => $url,
+      ]);
+      return $original_uri;
+    }
+
+    // Get the fallback filename.
+    $filename = $row->get($this->configuration['default_filename']) ?? 'file';
+
+    $mimeTypes = new MimeTypes();
+    // Find the proper fallback file extension if possible.
+    $extensions = $mimeTypes->getExtensions($type);
+    $extension = '';
+    if (!empty($extensions)) {
+      $extension = reset($extensions);
+      $filename .= '.' . $extension;
+    }
+    else {
+      $this->logger->notice("Couldn't determine file extension for %url", [
+        '%url' => $url,
+      ]);
+      return;
+    }
+    $this->logger->debug("Recognized extension as %ext", [
+      '%ext' => $extension,
+    ]);
```

Comment:
Will these debug log messages be kept? Maybe this one could be removed at least.

Reply:
@tadean Since this is an experimental module, we could leave the debug messages in for now if you think that would be useful. I think the "Sanitized to %filename" message would be worth keeping regardless. The other two debug messages don't seem as necessary to me, but whatever we decide is fine with me!